### PR TITLE
Iam 128

### PR DESCRIPTION
**Pull request by bencekov**
I'm aware of issues with the tests, however I think most of the implementation is ready to be reviewed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,14 +18,22 @@ juju deploy identity-platform-login-ui-operator
 To set the Ory Kratos endpoint to the UI server, do:
 
 ```console
-juju config identity-platform-login-ui-operator kratos_url={kratos_url:port}
+juju relate identity-platform-login-ui-operator kratos
```
Comment (nsklikas) [inline README.md:21]:
```suggestion
juju integrate identity-platform-login-ui-operator kratos
```
`juju relate` will be deprecated in favor of `juju integrate`

File: README.md [hunk 2]
```diff
@@ -18,14 +18,22 @@ juju deploy identity-platform-login-ui-operator
...
 ```
 
 To set the Ory Hydra endpoint to the UI server, do:
 
 ```console
-juju config identity-platform-login-ui-operator hydra_url={hydra_url:port}
+juju relate identity-platform-login-ui-operator hydra
```
Comment (nsklikas) [inline README.md:27]:
```suggestion
juju integrate identity-platform-login-ui-operator hydra
```

File: README.md [hunk 3]
```diff
@@ -18,14 +18,22 @@ juju deploy identity-platform-login-ui-operator
...
 ```
+
+To set the UI server's endpoint to Hydra and Kratos, do:
+
+```console
+juju relate hydra identity-platform-login-ui-operator
+juju relate kratos identity-platform-login-ui-operator
```
Comment (nsklikas) [inline README.md:34]:
```suggestion
juju integrate hydra identity-platform-login-ui-operator
juju integrate kratos identity-platform-login-ui-operator
```

File: metadata.yaml [hunk 4]
```diff
@@ -24,3 +24,15 @@ resources:
 requires:
   ingress:
     interface: ingress
+  endpoint-info:
```
Comment (nsklikas) [inline metadata.yaml:27]:
```suggestion
  hydra-endpoint-info:
```

File: src/charm.py [hunk 5]
```diff
@@ -79,10 +106,49 @@ def _update_pebble_layer(self, event: HookEvent) -> None:
...
         if self.unit.is_leader():
             logger.info("This app no longer has ingress")
+        self._update_pebble_layer(event)
+        self._update_login_ui_endpoint_relation_data(event)
+
+    def _update_login_ui_endpoint_relation_data(self, event: RelationEvent) -> None:
+        endpoint = (
+            self.ingress.url
+            if self.ingress.is_ready()
+            else f"{self.app.name}.{self.model.name}.svc.cluster.local:{APPLICATION_PORT}",
```
Comment (nsklikas) [inline src/charm.py:122]: I would suggest we provide no endpoint if ingress is not ready. The login ui needs to be accessible from the user and this domain name makes sense only within the k8s cluster.

File: requirements.txt [hunk 6]
```diff
@@ -1,4 +1,5 @@
 ops>=2.0.0
 lightkube
 lightkube-models
-jsonschema
+Jinja2
```
Comment (nsklikas) [inline requirements.txt:4]: Why is this needed?

File: tests/unit/conftest.py [hunk 7]
```diff
@@ -2,25 +2,66 @@
...
+    mocked_pebble_exec.return_value = mocked_process
+    yield mocked_pebble_exec
+
+
+@pytest.fixture()
+def mocked_pebble_exec_failed(mocked_pebble_exec: MagicMock) -> Generator:
+    mocked_pebble_exec.side_effect = ExecError(
+        exit_code=400, stderr="Failed to execute", stdout="Failed", command="test command"
+    )
+    yield
```
Comment (nsklikas) [inline tests/unit/conftest.py:67]: Nothing is returned here

File: tests/unit/test_charm.py [hunk 8]
```diff
@@ -78,12 +108,63 @@ def test_layer_updated(harness) -> None:
                 "command": "identity_platform_login_ui",
                 "startup": "enabled",
                 "environment": {
-                    "HYDRA_ADMIN_URL": TEST_HYDRA_URL,
-                    "KRATOS_PUBLIC_URL": TEST_KRATOS_URL,
+                    "HYDRA_ADMIN_URL": "http://hydra-public-url:80/testing-hydra",
+                    "KRATOS_PUBLIC_URL": "http://kratos-public-url:80/testing-kratos",
```
Comment (nsklikas) [inline tests/unit/test_charm.py:112]: Don't hard code this, get the data from the relation and match the pebble layer against those.

File: tests/unit/test_charm.py [hunk 9]
```diff
@@ -78,12 +108,63 @@ def test_layer_updated(harness) -> None:
...
                     "PORT": TEST_PORT,
                 },
             }
         },
     }
 
     assert harness.charm._login_ui_layer.to_dict() == expected_layer
+
+
+@pytest.mark.parametrize("port", ["8080"])
```
Comment (nsklikas) [inline tests/unit/test_charm.py:122]: No need for parametrizing in this test (you have only one value)

File: tox.ini [hunk 10]
```diff
@@ -9,13 +9,13 @@ envlist = fmt, lint, unit
 [vars]
 src_path = {toxinidir}/src/
 tst_path = {toxinidir}/tests/
-;lib_path = {toxinidir}/lib/charms/operator_name_with_underscores
-all_path = {[vars]src_path} {[vars]tst_path} 
+lib_path = {toxinidir}/lib/charms/
```
Comment (nsklikas) [inline tox.ini:12]: I think it's best to change this to 
```suggestion
lib_path = {toxinidir}/lib/charms/identity_platform_login_ui
```

This is only used for calculating coverage and I don't think we need to calculate our test coverage for libraries that belong to other charms.

File: README.md (hunk 2 above)
Comment (natalian98) [inline README.md:27]: I think you'll need to specify relation names when multiple integrations are in place for the 2 charms.
We could also simplify the `Interacting with Kratos and Hydra` part by splitting it into 2 parts - for Hydra and Kratos. For example:
### Hydra
This charm offers integration with `hydra-operator`. To integrate them, run:
```bash
juju integrate hydra:hydra-endpoint-info identity-platform-login-ui-operator:hydra-endpoint-info
juju integrate hydra:ui-endpoint-info identity-platform-login-ui-operator:ui-endpoint-info
```
### Kratos
This charm offers integration with `kratos-operator`. To integrate them, run:
```bash
juju integrate kratos:kratos-endpoint-info identity-platform-login-ui-operator:kratos-endpoint-info
juju integrate kratos:ui-endpoint-info identity-platform-login-ui-operator:ui-endpoint-info
```
What do you think?

File: config.yaml [hunk 11]
```diff
@@ -10,9 +10,4 @@ options:
       Acceptable values are: "info", "debug", "warning", "error" and "critical"
     default: "info"
     type: string
-  kratos_url:
```
Comment (natalian98) [inline config.yaml:13]: I think our PRs in this repo will be in conflict - I opened one for the kratos integration. I'm not sure if we want to add all relations in this PR or split them into different ones. @massigori what do you think?

File: lib/charms/identity_platform_login_ui/v0/login_ui_endpoint.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+import logging
+from typing import Dict, Optional
+
+from ops.charm import CharmBase, RelationCreatedEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+from ops.model import Application
+
+# The unique Charmhub library identifier, never change it
+LIBID = "placeholder_ID"
```
Comment (natalian98) [inline lib/charms/identity_platform_login_ui/v0/login_ui_endpoint.py:48]: You can register the library without publishing it yet. Refer to [this guide](https://juju.is/docs/sdk/charmcraft-create-lib) to get `LIBID`

File: src/charm.py [hunk 13]
```diff
@@ -32,6 +40,8 @@ def __init__(self, *args):
         super().__init__(*args)
         self._container_name = "login-ui"
         self._container = self.unit.get_container(self._container_name)
+        self._hydra_relation_name = "endpoint-info"
```
Comment (natalian98) [inline src/charm.py:43]: We'll probably change this name to `hydra-endpoint-info` to be consistent, leaving a comment to keep track of it

File: tests/integration/test_charm.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+async def test_has_ingress(ops_test: OpsTest):
+    # Get the traefik address and try to reach kratos
+    public_address = await get_unit_address(ops_test, TRAEFIK_PUBLIC_APP, 0)
+
+    resp = requests.get(
+        f"http://{public_address}/{ops_test.model.name}-{APP_NAME}/api/self-service/errors?id=1111"
+    )
+
+    assert resp.status_code == 404
```
Comment (natalian98) [inline tests/integration/test_charm.py:75]: Is `test_has_ingress` a good name for this test? It tries to reach an address under traefik, but then asserts that the resource wasn't found - I suppose you are checking that there were no errors?

**Review by natalian98 — COMMENTED** [review]
Could you also add a workflow to automatically publish `login_ui_endpoint.py` library if changed on push? Check [this one](https://github.com/canonical/hydra-operator/blob/main/.github/workflows/publish-lib.yaml) for example

File: lib/charms/identity_platform_login_ui/v0/login_ui_endpoint.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    def send_endpoint_relation_data(self, endpoint: str) -> None:
+        """Updates relation with endpoint info."""
+        if not self._charm.unit.is_leader():
+            return
+
+        relations = self.model.relations[self._relation_name]
+        for relation in relations:
+            relation.data[self._charm.app].update(
+                {
+                    "endpoint": endpoint,
```
Comment (natalian98) [inline lib/charms/identity_platform_login_ui/v0/login_ui_endpoint.py:100]: Do we want to provide only the bare address of login ui without specifying its full endpoints?
Hydra and Kratos need the login form and error urls. Right now both construct these endpoints by adding `/login` or `/oidc_error` to the login ui url (see [here](https://github.com/canonical/kratos-operator/blob/main/src/charm.py#L165)), but as `identity-platform-login-ui-operator` implements them, I think it should be able to provide its full endpoints.

Edit: The following urls should be placed in the databag (if implemented): consent, error, login, registration, default browser url (which is what you currently have).

File: tests/integration/test_charm.py (hunk 14 above)
Comment (bencekov) [reply tests/integration/test_charm.py:75]: I agree it's not a great test. Basically I try to interact with the Ory API without registering a flow, or doing complicated things. I'll change the test, this was misguided.
